### ISSUE TITLE
Distributed training optimization

### ISSUE THREAD

**Issue by JianbangZ** [tensorflow/tensorflow#4465]
After playing with the current distributed training implementation for a while, I think it views each GPU as a separate worker.However, It is common now to have 2~4 GPUs in one box. Isn't it better to adopt the single box multi-GPU methodology to compute average gradients in single box first and then sync up across multiple nodes? This way it ease the I/O traffic a lot, which is always the bottleneck in data parallelism.

**Comment by vrv** [comment]
I think the example implementations might be one GPU per worker, but nothing prevents you from defining a worker having 8 GPUs, replicating the model across each GPU in a worker, averaging gradients before sending it to parameter servers.  TF is low-level enough that these are all possible.  To make the system easy to understand, the examples are simpler than what is possible.

I suspect higher-level libraries will be added over time that canonicalizes these types of model deployments that are becoming more popular.
